src/monitor.cpp

```diff
@@ -160,108 +160,208 @@ complex<double> fields_chunk::get_field(component c, const ivec &iloc) const {
+  Vinv[1 + 3*0] = 1/det * (V[1 + 3*2]*V[2 + 3*0] - V[1 + 3*0]*V[2 + 3*2]);
+  Vinv[1 + 3*1] = 1/det * (V[0 + 3*0]*V[2 + 3*2] - V[0 + 3*2]*V[2 + 3*0]);
+  Vinv[1 + 3*2] = 1/det * (V[0 + 3*2]*V[1 + 3*0] - V[0 + 3*0]*V[1 + 3*2]);
+  Vinv[2 + 3*0] = 1/det * (V[1 + 3*0]*V[2 + 3*1] - V[1 + 3*1]*V[2 + 3*0]);
+  Vinv[2 + 3*1] = 1/det * (V[0 + 3*1]*V[2 + 3*0] - V[0 + 3*0]*V[2 + 3*1]);
+  Vinv[2 + 3*2] = 1/det * (V[0 + 3*0]*V[1 + 3*1] - V[0 + 3*1]*V[1 + 3*0]);
 }
 
-double structure::get_chi1inv(component c, direction d, const vec &loc, bool parallel) const {
+double structure_chunk::get_chi1inv_at_pt(component c, direction d, int idx, double omega) const {
+  double res = 0.0;
+  if (is_mine()){
+    // ----------------------------------------------------------------- //
+    // ---- Step 1: Get instantaneous chi1 tensor ----------------------
+    // ----------------------------------------------------------------- //
+    
+    int my_stuff = E_stuff;
+    component comp_list[3];
+    if (is_electric(c)) {
+      comp_list[0] = Ex; comp_list[1] = Ey; comp_list[2] = Ez;
+      my_stuff = E_stuff;
+    }else if (is_magnetic(c)) {
+      comp_list[0] = Hx; comp_list[1] = Hy; comp_list[2] = Hz;
+      my_stuff = H_stuff;
+    } else if (is_D(c)) {
+      comp_list[0] = Dx; comp_list[1] = Dy; comp_list[2] = Dz;
+      my_stuff = D_stuff;
+    } else if (is_B(c)) {
+      comp_list[0] = Bx; comp_list[1] = By; comp_list[2] = Bz;
+      my_stuff = B_stuff;
+    }
+
+    std::vector<double> chi1_inv_tensor(9,0);
+    std::vector<double> chi1_tensor(9,0);
+
+    // Set up the chi1inv tensor
+    for (int com_it=0; com_it<3;com_it++){
+      for (int dir_int=0;dir_int<3;dir_int++){
+        if (chi1inv[comp_list[com_it]][dir_int] )
+          chi1_inv_tensor[com_it + 3*dir_int] = chi1inv[comp_list[com_it]][dir_int][idx];
+        else if(dir_int == component_direction(comp_list[com_it]))
+          chi1_inv_tensor[com_it + 3*dir_int] = 1;
+        else
+          chi1_inv_tensor[com_it + 3*dir_int] = 0;
+      }
+    }
+    
+    
+    matrix_invert(chi1_tensor, chi1_inv_tensor); // We have the inverse, so let's invert it.
+    
+    // ----------------------------------------------------------------- //
+    // ---- Step 2: Evaluate susceptibilities of each tensor element ---
+    // ----------------------------------------------------------------- //
+    
+    // loop over tensor elements
+    for (int com_it=0; com_it<3;com_it++){
+      for (int dir_int=0;dir_int<3;dir_int++){
+        std::complex<double> eps(chi1_tensor[com_it  + 3*dir_int],0.0);
+        component cc = comp_list[com_it];
+        direction dd = (direction)dir_int;
+        // Loop through and add up susceptibility contributions
+        // locate correct susceptibility list
+        susceptibility *my_sus = chiP[my_stuff];
+        while (my_sus) {
+          if (my_sus->sigma[cc][dd]) {
+            double sigma = my_sus->sigma[cc][dd][idx];
```

Comment:
This is a little worrisome — doesn't this correspond to different points in the Yee grid for the same `idx`?   Seems like you have to do some averaging with neighboring values like in h5 output and array slices.